### DISCUSSION TITLE
RX real time thread

### DISCUSSION THREAD

**Discussion by cspiel1** [baresip/baresip#2445] [Show and tell]
This discussion describes the SW architecture and necessary steps to move the whole audio/(video) receive (RX) processing into a separate thread.

### Current SW 
![current-rx](https://user-images.githubusercontent.com/14180877/217498682-ecec34cb-0e23-4062-9799-e61a2a282df0.png)
Architecture

The main thread performs the processing of
- all network sockets (not only RX audio),
- all SW timers
- all user/application interaction (e.g. module menu)
- RTCP/RTP metrics computation, queries and sending RTCP packets
- start/stop and changes of the RTP RX audio processing
- logging (writing to stdout)
- passing events to the application (dbus, tcp sockets)
- sending of TX audio
- video RX/TX

These all might lead to long blocking function calls and thus to SW jitter and to extra audio latency in the RX pipeline.

### Planned SW Architecture
We replace the Main Thread for the audio RX processing by a separate RX thread.

![planned-rx](https://user-images.githubusercontent.com/14180877/217498599-5223924d-bc6e-42d4-b150-3802e371c033.png)

- The RTP audio socket should have its own polling loop that is processed in the RX thread.
- The most difficult part is to make all the data exchange with the Main Thread thread safe. We have to specify a minimal set of tasks that have to be processed in the RX thread.
- We want to specify a minimal set of data that are owned by the RX thread and are read or changed by the main thread very seldom. These data fields have to be made thread safe. And the main thread may not lock these data fields longer than for a simple read or copy.
- Other data structures that are a result in the RX thread and imply a longer processing in the main thread have to be passed as duplicates via a queue to the main thread. (e.g. with `re_thread_async_main`)

### Necessary Steps

There are two involved source files, that are touched by the audio RX thread.

- audio.c
- stream.c

Details will follow.

**Comment by cspiel1** [comment]
## stream.c
### Data structures
- [x]  The struct `receiver` is moved to a separate receiver.c
  - [x ] A mutex protects fields that are accessed from both threads.
- [ ] No other data has to be used from RX thread, except:
  - [ ] the RTP socket `stream->rtp` is used for ???
    - [ ] helpers
  - [x] SDP data is not accessed from RX thread
  - [x] stream_destructor() : Be sure that the RX thread is stopped gracefully before essential other objects are de-referenced.

### Handler called from RX thread
- rtph - RTP receive handler
- [x] pth - Payload type handler (but not everything in the payload type handler should be done in RX thread)
  - [x] The decode may be changed from `stream_pt_handler`. --> PR:
    - `mem_deref(rx->auplay)` could be done also from main thread
    - `mem_deref(rx->dec)` same
    - even `audio_decoder_set()` is called from main thread normally
    - Thus the function `audio_decoder_set()` should be called only from main thread.
  - [x] In-bound DTMF processing --> in main thread (https://github.com/baresip/baresip/pull/2444)
### Handler that should not be called from RX thread
We have to pass such calls via an mqueue or an asynchronous worker.
- [x] rtcph - RTCP handler `stream_rtcp_h` --> video.c
- [x] sessrtcph - session rtcp handler `stream_rtcp_h` → call.c, mediatrack.c
- [x] rtpestabh - incoming RTP established handler `stream_rtpestab_h` → call.c, mediatrack.c
### Timers
- [x]  NAT pinhole timer tmr_natph.
  - [x] The tmr_cancel has to be called from main thread, or
  - [x] `tmr_cancel()` now is thread safe.

### UDP Helpers
There are a plenty of UDP helpers registered (e.g. mnat, menc). Each received RTP packet is also passed to the UDP helpers.

- [x] Thread Safe tmr handling -> https://github.com/baresip/re/pull/690
- [x] mnat - stun, ice, turn have a recev-handler and looks in the RTP stream for STUN responses. So might call the MNAT connected handler in stream.c.
  - [x] DNS for stun -> starts timers
  - [x] The helper (e.g. stun, ice, turn) need to be thread safe.
  - [x] The timers are thread safe.
- [x] menc - does SRTP decryption which should be done in the RX thread
  - [x] allocated media encryption media state `mes` is not changed in main thread
  - [x] dereferencing of `mes` effects in an unlink of the helper -->  https://github.com/baresip/re/pull/732
  - [x] start of media encoding is thread safe, because the RX thread is started after MENC is initialized completely with function `stream_start_mediaenc()`. It is possible that the first few RTP packets are handled in main thread and then the control is moved to RX thread.
  - [x] `srtp` is thread safe
    - [x] media session `mencs` of stream --> only used in main thread
    - [x] `struct sdp_media` of stream --> only used in main thread
    - [x] stopping is thread safe, because `sdpm` and `crypto_suite` that are de-referenced before the `uh_rtp` and `uh_rtcp` are not used from RX thread
  - [x] `dtls_srtp` is thread safe
    - [x] media session `mencs` of stream --> only used in DTLS handshake, done in main thread
    - [x] stopping is thread safe because the UDP helper `uh_srtp` is de-referenced first in the destructor
    - [x] `struct sdp_media` of stream --> only used in main thread
  - [x] `gzrtp` is thread safe
    - [x] `media_sess` is only used in main thread
    - [x] stopping is thread safe --> The flag `m_started` is not enough. The helper has to be removed first in the destructor. https://github.com/baresip/baresip/pull/2492
    - [x] `struct sdp_media` of stream --> only used in main thread
  - [x] menc is thread safe in general --> no it is not. When writing a `menc` module it has to be taken care, that:
    - [x] The `menc_sess` should be used only during initialization or in callback function that are called from main thread.
    - [x] The `sdp_media` should be used only during initialization.
    - [x] The `struct stream` is not used from RX thread.
- [x] bundle
  - `bundle_start_socket()` called from main thread passed the socket and a the stream list.
  - it registers a UDP recv helper for the "base" stream (`bundle.c` function `udp_helper_recv_handler()`) that calls
    - [x] `lookup_remote_ssrc()` which
       - [x] searches the stream list `streaml` --> ok: the streams are allocated before the bundle
       - [x] calls thread safe `rx_get_ssrc()` in `receiver.c` to find the correct `stream`
       - [x] sanitizer warning: `stream_destructor()` calls `list_unlink()` while rx-thread calls `lookup_remote_ssrc()` -> solve by deref first the receiver. This stops the RX thread before the stream is removed from its list.
    - [x] it uses the streams bundle and calls following helpers and the receive handler of the streams socket
      - The helper list may be changed by received SDP offer/answer from main thread.
      - The helpers are called from RX thread.
      - [x] ensure that the helper list does not change during processing a packet --> https://github.com/baresip/re/pull/732

**Comment by cspiel1** [comment]
## audio.c --> separated aureceiver.c
### Data structures
- [x] `struct aurpipe` holds not only data for the RX thread
  - [x] `struct audec_state *dec`: The decoder is used for each audio frame to decode it.
    - [x] `aup_decoder_set()` changes the decoder. It has to be made thread safe.
  - [x] `rx→ap` a pointer to the auplay used only in main thread
  - [x] `rx→auplay_prm` only set during start of audio and read in main thread
  - [x] `rx→ac` only set during start and when the codec changes
    - [x] `aup_decoder_set()` made thread safe.
    - [x] print, debug functions made thread safe
    - [x] getter and setter both are called from main thread 
  - [x] `aup→aubuf`:  aubuf is thread safe
  - [x] `aup→ssrc`: rw in RX thread for each frame (exclusively)
  - [x] `aup→filtl`: Set only initially in mainthread. Then used for each audio frame. May be changed seldom in main thread.
    - [x] The filter list itself is thread safe.
    - [x] The `dech()` function don't share any data with main thread.
  - [x] `rx→module` and `rx→device`: Has nothing to do with the RX thread.
  - [x] `aup→sampv`: Allocated at start in main thread. Then used for each frame in the RX thread.
  - [x ] `rx→ptime`: Has nothing to do with the RX thread.
  - [x] `rx→pt`: Is read/set for each audio frame in the `stream_pt_handler()`. If it changes, the work is passed to the main thread.
  - [x] `rx→level_last` and `rx->level_set`: Is set in the RX thread for each frame. And can be read from main thread. --> thread safe by mutex
  - [x] `rx→play_fmt`: has nothing to do with RX thread
  - [x] `rx→ts_recv`: written/read for each audio frame in RX thread. Read very seldom in main thread. --> thread safe
  - [x] `rx→stats`: written in RX thread. Read very seldom in main thread. --> thread safe
  - [x] Solution:
    - [x] Add a sub-struct for data that is changed for each audio frame.
    - [x] The data fields in this sub-struct that are also used in other thread, have to be made thread safe.

- [x] Other data structures and fields should only be used with read access and made thread safe by
  - [x] be sure that the data field does not change during RX thread is running, or
  - [x] use atomic or a mutex.
  - [x] A mutex should be locked in RX thread only very short for making a copy or a single read of a primitive type. Debug functions should write in mbuf while lock is held.

- [x] Changes of the audio filter list, the decoder
  - [x] made thread safe
- [x] Changes of the auplay should not need a mutex lock of the RX thread

### Handler called in audio.c

- [x] eventh (DTMF) have to be called from main thread. It has to be passed via queue to the main thread. --> See `async_work_main()`!
- [x] levelh - audio level handler has nothing to do with the RX thread
- [x] errh - audio error handler is called from audio source thread. Has nothing to do with the RX thread.
### Timers
- [ ] There are no timers in audio.c or called from indirectly from audio.c, really?

**Comment by sreimers** [comment]
I think a good starting point could be a experimental RX_THREAD config option. So we can test/verify with clang/gcc thread sanitizer which runtime race conditions are still happening. with good test coverage we should also be able to ensure that no new race condition bugs are introduced in the future.

**Reply by cspiel1** [reply]
okay. This will help to verify against race conditions. But lets solve first the concurrency problems we already know!

**Comment by cspiel1** [comment]
@sreimers Should we make a feature branch with similar PR review and github workflows?

We have already a PR for payload type handler and now I plan to create a second one. There would be already a conflict.

**Reply by sreimers** [reply]
Maybe easier to combine payload type changes in https://github.com/baresip/baresip/pull/2444 with separate commits for now. release day is next week.

**Comment by cspiel1** [comment]
I try it with:
https://github.com/baresip/baresip/tree/audio_rxthread

Then will make PRs with merge to `audio_rxthread` and check if the auto test are started.

**Comment by cspiel1** [comment]
I don't understand why there are two RTCP handlers.

```c
	if (s->rtcph)
		s->rtcph(s, msg, s->arg);

	if (s->sessrtcph)
		s->sessrtcph(s, msg, s->sess_arg);
```

If `call.c` needs to be informed about an RTCP packets, then audio.c and video.c should have its own function pointer.

**Reply by cspiel1** [reply]
For now I will take it like it is. So receiver.c, will have an `rtcph` and an `sessrtcph` with different arguments. We need the `sess_arg` for other callback functions anyway.

**Comment by cspiel1** [comment]
@sreimers, @alfredh What should we do with getters that have to lock a mutex?

`mtx_lock()`  doesn't have a const parameter. But the getters should be const.

**Reply by sreimers** [reply]
If heap pointer from `mutex_alloc` is used, const for getter struct should be fine I think.

**Comment by cspiel1** [comment]
With commit https://github.com/baresip/baresip/pull/2454/commits/44d35a95d39631be51cfc382aaed81155a59ec7a the MNAT connect processing is passed to the main thread.

Other stuff done in the UDP helper for mnat will be processed in the RX thread, like DNS discovery and STUN timers. @sreimers both should be safe now, right?

If so it might lead to SW jitter. But this should arise only during startup of the stream. For now we should accept this. Maybe try to improve this later.

**Reply by sreimers** [reply]
At first look it should be fine, `mnat->sessh`, `mnat->mediah` and `mnat->updateh` are called from main thread, right?
DNS handling is currently only needed for `mnat` session allocation I think. So yes it looks good to me.

**Reply by cspiel1** [reply]
Yes, I think so. All three are called from main thread.

**Comment by cspiel1** [comment]
Next point is `menc`. The UDP helper for `menc` run in RX thread and media decryption should be done in RX thread without any interaction with the main thread. So we are also safe here, right?

**Reply by cspiel1** [reply]
Still not sure. Have to look closer into modules `srtp` and `dtls_srtp`.

**Comment by cspiel1** [comment]
### Mutex of `aurpipe`
regarding: https://github.com/baresip/baresip/pull/2454/commits/8d847a91eb66e89c0f9a537aaea4d415099d4114

The mutex of struct `aurpipe` is hold by the RX-thread while one frame is decoded. On the other side it is hold by main and auplay threads very short to read/write fields from `aurpipe`. This is okay if the called `aurpipe` functions are called very rarely. 

If a function that accesses `aurpipe` is called frequently, it should not lock the mutex and use e.g. atomic fields instead.
E.g. `stats.latency` is computed in the RX-thread and can be read in other threads by accessing the atomic field. The function `aup_latency()` is called indirectly by a timer every 500ms of module `vumeter`.

E.g. `aup_debug()` is called by the user. Hopefully not to often. A call to `aup_debug()` may block during the decoding of one audio frame. Thus the main loop is blocked also for at most one audio frame.

**Comment by cspiel1** [comment]
### Update
Refactoring was done in two steps:
- 33047e60cf14324603d2effb79a1b85140557b17
- b9dbb8978ed6962a75f4bb59b464cdb5fef41272

A final PR should activate the RX thread: https://github.com/baresip/baresip/pull/2828

### video / bundle
Thread safety for video processing in RX thread currently can not be guaranteed. Thus RX thread is disabled for video and for bundle mode.

### RTCP
In baresip the RTCP call back handler passes the work to the main thread.
But the structs in `re` for rtcp are still written in RX thread while read from main thread. This has to be solved.
